### PR TITLE
[inductor] Prevent reusing aliased buffers if aliases still have uses

### LINKED ISSUE #100314: Inductor miscompilation involving reuse of cat inputs
### 🐛 Describe the bug

Repro: https://gist.github.com/ezyang/820feb908b3a6397bf0799bfd0ad6568
Generated inductor code: https://gist.github.com/ezyang/aab4f2ae268d2a2bf68f213ad3f9aac9

Analyze on the repro says

```
Saving inductor intermediates: 100%|███████████████████████████████████████████████████████| 4/4 [00:03<00:00,  1.29it/s]
NONDETERMINISTIC INDUCTOR at add (4047fc0a58494c7a5922c7bdfcdda3c9475288d6 != a4b72fbe3f71371660a4de8dba8ec4f121f96587)  
Checking inductor determinism: 100%|███████████████████████████████████████████████████████| 4/4 [00:00<00:00, 93.07it/s]
Saving float64 intermediates: 100%|████████████████████████████████████████████████████████| 4/4 [00:00<00:00, 67.64it/s]
Checking float64 determinism: 100%|████████████████████████████████████████████████████████| 4/4 [00:00<00:00, 70.44it/s]
DIVERGED at add_2: RMSE (res-fp64): 0.45063, (ref-fp64): 0.00000 and shape=torch.Size([8, 1, 768])                       
Checking divergence: 100%|█████████████████████████████████████████████████████████████████| 4/4 [00:00<00:00, 47.21it/s]
Loading inputs: 100%|████████████████████████████████████████████████████████████████████| 28/28 [00:09<00:00,  2.82it/s]

```

EDIT: the nondeterminism is fake, it's due to only part of an empty buffer getting filled.

EDIT 2: the names in FX IR are misleading because when we retrace all the variables get renamed. Here are the actual names

```
    def forward(self, arg0_1: f32[768], arg1_1: f32[768], arg2_1: f32[768], arg3_1: f32[768], arg4_1: f32[768], arg5_1: f32[768], arg6_1: f32[8, 576, 768], arg7_1: f32[8, 1, 768], arg8_1: f32[8, 768], arg9_1: f32[8, 577, 1], arg10_1: f32[8, 577, 1], arg11_1: f32[768, 768], arg12_1: f32[128, 1, 577], arg13_1: f32[768, 768]):
        # No stacktrace found for following nodes
        view: f32[8, 1, 768] = torch.ops.aten.view.default(arg8_1, [8, 1, 768]);  arg8_1 = None
        mul: f32[8, 1, 768] = torch.ops.aten.mul.Tensor(arg0_1, view);  arg0_1 = view = None
        add: f32[8, 1, 768] = torch.ops.aten.add.Tensor(arg7_1, mul);  arg7_1 = mul = None
        cat: f32[8, 577, 768] = torch.ops.aten.cat.default([add, arg6_1], 1);  arg6_1 = None
        sub: f32[8, 577, 768] = torch.ops.aten.sub.Tensor(cat, arg9_1);  cat = arg9_1 = None
        mul_1: f32[8, 577, 768] = torch.ops.aten.mul.Tensor(sub, arg10_1);  sub = arg10_1 = None
        mul_2: f32[8, 577, 768] = torch.ops.aten.mul.Tensor(mul_1, arg2_1);  mul_1 = arg2_1 = None
        add_1: f32[8, 577, 768] = torch.ops.aten.add.Tensor(mul_2, arg3_1);  mul_2 = arg3_1 = None
        view_1: f32[4616, 768] = torch.ops.aten.view.default(add_1, [4616, 768]);  add_1 = None
        addmm: f32[4616, 768] = torch.ops.aten.addmm.default(arg4_1, view_1, arg11_1);  arg4_1 = view_1 = arg11_1 = None
        view_2: f32[8, 577, 768] = torch.ops.aten.view.default(addmm, [8, 577, 768]);  addmm = None
        view_3: f32[8, 577, 16, 48] = torch.ops.aten.view.default(view_2, [8, 577, 16, 48]);  view_2 = None
        permute: f32[8, 16, 577, 48] = torch.ops.aten.permute.default(view_3, [0, 2, 1, 3]);  view_3 = None
        expand: f32[8, 16, 577, 48] = torch.ops.aten.expand.default(permute, [8, 16, 577, 48]);  permute = None
        clone: f32[8, 16, 577, 48] = torch.ops.aten.clone.default(expand, memory_format = torch.contiguous_format);  expand = None
        view_4: f32[128, 577, 48] = torch.ops.aten.view.default(clone, [128, 577, 48]);  clone = None
        bmm: f32[128, 1, 48] = torch.ops.aten.bmm.default(arg12_1, view_4);  arg12_1 = view_4 = None
        view_5: f32[8, 16, 1, 48] = torch.ops.aten.view.default(bmm, [8, 16, 1, 48]);  bmm = None
        permute_1: f32[8, 1, 16, 48] = torch.ops.aten.permute.default(view_5, [0, 2, 1, 3]);  view_5 = None
        view_6: f32[8, 1, 768] = torch.ops.aten.view.default(permute_1, [8, 1, 768]);  permute_1 = None
        view_7: f32[8, 768] = torch.ops.aten.view.default(view_6, [8, 768]);  view_6 = None
        addmm_1: f32[8, 768] = torch.ops.aten.addmm.default(arg5_1, view_7, arg13_1);  arg5_1 = view_7 = arg13_1 = None
        view_8: f32[8, 1, 768] = torch.ops.aten.view.default(addmm_1, [8, 1, 768]);  addmm_1 = None
        mul_3: f32[8, 1, 768] = torch.ops.aten.mul.Tensor(arg1_1, view_8);  arg1_1 = view_8 = None
        add_2: f32[8, 1, 768] = torch.ops.aten.add.Tensor(add, mul_3);  add = mul_3 = None
        return (add_2,)
```

### Versions

master

cc @gchanan @zou3519 @soumith @msaroufim @wconstab @ngimel @bdhirsh @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### PR DESCRIPTION
**Pull request by ngimel**
Fixes #100314
In dependencies, we should track not only immediately used buffer, but also aliased buffers that point to it, otherwise we can reuse and overwrite the buffer while there are still pending uses. 

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Comment by ezyang** [comment]
Is there some sort of debug mode assert we could also add to make it easier to localize the problem here? At some point I did try `inplace_buffers = False` which Horace thought might have been the problem, but it didn't seem like that fixed this issue.

**Comment by ngimel** [comment]
Not currently, but I can add a config for that. 
Q: can we have nested aliased layouts? If so, what I've added is not sufficient

File: torch/_inductor/scheduler.py [hunk 1]
```diff
@@ -519,6 +531,12 @@ def debug_str_extra(self):
     def used_buffer_names(self) -> Set[str]:
         return functools.reduce(set.union, [x.used_buffer_names() for x in self.snodes])
 
+    @cache_on_self
+    def used_or_aliased_buffer_names(self) -> Set[str]:
+        return functools.reduce(
+            set.union, [x.used_or_aliased_buffer_names() for x in self.snodes]
```
Comment (Skylion007) [inline torch/_inductor/scheduler.py:537]:
```suggestion
            set.union, (x.used_or_aliased_buffer_names() for x in self.snodes)
```
This reduction can use a generator.

**Review by jansel — APPROVED** [review]

**Comment by ngimel** [comment]
@ezyang I've added config to disable buffer reuse, and looks like nested AliasedLayouts are not created, so what's here should be enough.

**Pull request closed without merging**